pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/KafkaProducer.java

```diff
@@ -0,0 +1,398 @@
+        configs, keySerializer, valueSerializer)), keySerializer, valueSerializer);
+  }
+
+  public KafkaProducer(Properties properties) {
+    this(new ProducerConfig(properties), null, null);
+  }
+
+  public KafkaProducer(Properties properties, Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+    this(new ProducerConfig(ProducerConfig.addSerializerToConfig(
+        properties, keySerializer, valueSerializer)), keySerializer, valueSerializer);
+  }
+
+  //TODO: Kafka's implementation throws a KafkaException upon failure, it doesn't apply here.
+
+  @VisibleForTesting
+  @SuppressWarnings("unchecked")
+  public KafkaProducer(ProducerConfig configs, Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+
+    producerConfig = configs;
+
+    project = configs.getString(ProducerConfig.PROJECT_CONFIG);
+
+    if (keySerializer == null) {
+      this.keySerializer = configs.getConfiguredInstance(
+          ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, Serializer.class);
+
+      this.keySerializer.configure(configs.originals(), true);
+    } else {
+      configs.ignore(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG);
+      this.keySerializer = keySerializer;
+    }
+
+    if (valueSerializer == null) {
+      this.valueSerializer = configs.getConfiguredInstance(
+          ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
+
+      this.valueSerializer.configure(configs.originals(), false);
+    } else {
+      configs.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
+      this.valueSerializer = valueSerializer;
+    }
+
+    closed = new AtomicBoolean(false);
+    retries = configs.getInt(ProducerConfig.RETRIES_CONFIG);
+    timeout = configs.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+    retriesMs = configs.getLong(ProducerConfig.RETRY_BACKOFF_MS_CONFIG);
+    lingerMs = configs.getLong(ProducerConfig.LINGER_MS_CONFIG);
+    batchSize = configs.getInt(ProducerConfig.BATCH_SIZE_CONFIG);
+    autoCreate = configs.getBoolean(ProducerConfig.AUTO_CREATE_CONFIG);
+    elementCount = configs.getLong(ProducerConfig.ELEMENTS_COUNT_CONFIG);
+    isAcks = configs.getString(ProducerConfig.ACKS_CONFIG).matches("(-)?1|all");
+    bufferMemorySize = configs.getLong(ProducerConfig.BUFFER_MEMORY_CONFIG);
+
+    String Id = configs.getString(ProducerConfig.CLIENT_ID_CONFIG);
+    if (Id.length() <= 0)
+      clientId = "producer-" + CLIENT_ID.getAndIncrement();
+    else 
+      clientId = Id;
+
+    configs.originals().put(ProducerConfig.CLIENT_ID_CONFIG, clientId);
+    List<ProducerInterceptor<K, V>> interceptorList =
+        (List) (new ProducerConfig(configs.originals())).getConfiguredInstances(
+            ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, ProducerInterceptor.class);
+    this.interceptors =
+        interceptorList.isEmpty() ? null : new ProducerInterceptors<>(interceptorList);
+    
+    publishers = new ConcurrentHashMap<>();
+  }
+
+  //TODO: deal with these magic numbers.
```

Comment:
The " + 1 ", " + 2 " thing, nevermind I think they are good.
It's a TODO for me to check later.